src/compiler/checker.ts

```diff
@@ -8822,6 +8959,177 @@ module ts {
+                return (<GenericType>promise).typeArguments[0];
+            }
+            
+            let globalPromiseLikeType = getInstantiatedGlobalPromiseLikeType();
+            if (globalPromiseLikeType === emptyObjectType || !isTypeAssignableTo(promise, globalPromiseLikeType)) {
+                return undefined;
+            }
+
+            let thenFunction = getTypeOfPropertyOfType(promise, "then");
+            if (thenFunction && allConstituentTypesHaveKind(thenFunction, TypeFlags.Any)) {
+                return undefined;
+            }
+            
+            let thenSignatures = thenFunction ? getSignaturesOfType(thenFunction, SignatureKind.Call) : emptyArray;
+            if (thenSignatures.length === 0) {
+                return undefined;
+            }
+            
+            let onfulfilledParameterType = getUnionType(map(thenSignatures, getTypeOfFirstParameterOfSignature));
+            if (allConstituentTypesHaveKind(onfulfilledParameterType, TypeFlags.Any)) {
+                return undefined;
+            }
+            
+            let onfulfilledParameterSignatures = getSignaturesOfType(onfulfilledParameterType, SignatureKind.Call);
+            if (onfulfilledParameterSignatures.length === 0) {
+                return undefined;
+            }
+            
+            let valueParameterType = getUnionType(map(onfulfilledParameterSignatures, getTypeOfFirstParameterOfSignature));
+            return valueParameterType;
+        }
+        
+        function getTypeOfFirstParameterOfSignature(signature: Signature) {
+            return getTypeAtPosition(signature, 0);
+        }
+        
+        /**
+          * Gets the "awaited type" of a type.
+          * @param type The type to await.
+          * @remarks The "awaited type" of an expression is its "promised type" if the expression is a 
+          * Promise-like type; otherwise, it is the type of the expression. This is used to reflect
+          * The runtime behavior of the `await` keyword.
+          */
+        function getAwaitedType(type: Type, location?: Node): Type {
+            // `seen` keeps track of types we've tried to await to avoid cycles.
+            // This is to protect against a bad actor with a mutually recursive promised type:
+            //
+            //  declare class PromiseA { 
+            //      then(onfulfilled: (value: PromiseB) => any, onrejected?); 
+            //  }
+            //  declare class PromiseB { 
+            //      then(onfulfilled: (value: PromiseA) => any, onrejected?); 
+            //  }
+            // 
+            let seen: boolean[];
+            while (true) {
+                let promisedType = getPromisedType(type);
+                if (!promisedType) {
+                    // this type could not be unwrapped further. We need to check to
+                    // ensure it is not a thenable
+                    if (checkNonThenableType(type, location)) {

```

Comment:
This is due to how `await` has been specified for ES7, and how Promise resolution works. The operand to `await` can either be a non-promise value (e.g. `await 1`), or a promise value. So if you `await` a `number`, the type is `number`. If you `await` a `Promise<number>`, the type is `number`. If you `await` an object with a callable `then` property, that does not have a signature compatible with a Promise, we report an error.

So, if `getPromisedType(type)` is `undefined`, then we would return `type`, unless `type` is a non-promise "thenable".